### PR TITLE
Added JS support

### PR DESCRIPTION
**Pull request by JohnAD**
This is a work in progress, please do not merge until the `[WIP]` prefix has been removed.

This is my second pass at adding Javascript (JS) support. This attempt will properly use the `TerminalBuffer` storage.

This first commit is simply a refactoring to prepare for the new code. In fact, it temporarily breaks windows support as I've only refactored for posix so far. I'll correct that fairly soon.

- [x] refactor generally to separate target interests
- [x] refactor/test posix
- [x] refactor/test windows
- [x] add `timerLoop(ms)` for cross-platform help
- [x] JS: display text and refresh
- [x] JS: color support
- [x] JS: Style support
- [x] JS: keyscanning
- [x] JS: (optional) add simple text input
- [x] JS: blinking cursor support
- [x] JS: add documentation

### CODE REVIEW

**Comment by JohnAD** [comment]
I brought a windows laptop home for the weekend. I'll get windows compilation working again soon; probably tomorrow. So far, everything is going very cleanly.

There is a odd problem with cursor positioning in posix. The cursor moves to the right spot from buffer point of view. But the visible cursor seems to be "off by one" sometimes. I'll try to fix that also.

**Comment by JohnAD** [comment]
Style is mostly working. All that is left for me to do is fix up `styleReverse` and update the documentation.

image

vs what is seen in a bash terminal:

image

**Comment by JohnAD** [comment]
Almost done. I need to document a few extra things. And, I will still want to fix that `showCursor` bug that sometimes appears.

**Comment by JohnAD** [comment]
Done! This PR is ready for review. Thanks!

**Comment by JohnAD** [comment]
I added a example program called "badtictactoe.nim" to demonstrate the `timerLoop` and user-input routines more fully.

As seen in a posix terminal:

image

As seen in a web broswer:

image

**Comment by enthus1ast** [comment]
This looks awesome @JohnAD !
I'm eager to try it!
Is mouse supported?

**Comment by JohnAD** [comment]
Thanks!

Mouse support is not in yet. There are, in fact, two more things I'd like to add:

1. Mouse support; tested against the main broswers (Chrome/Firefox/Edge/Safari/etc.)
2. A mechanism for "Enter" to work. By default, a web browser treats a user pressing enter as a message to "submit the form". So, the Enter key is never seen. One work-around is using Ctrl-Enter instead. That already works. (Reminder to self: document this.) But another method is to build a javascript method of translating a client-side form submission has an Enter key. I've got some experimenting to do.

But, I'm refraining for now. I don't want to add further complication without this first PR being reviewed/accepted. I'd rather not build further on a structure that might need to be changed on review.

**Comment by JohnAD** [comment]
(BTW, I probably won't get to it this year, but I'd like to also add support for Android in-screen terminals via the nim Godot bindings. That usage would also be event-driven like JS. But, one thing at a time!)

**Comment by JohnAD** [comment]
Does anyone have a method of contacting @johnnovak ? I've not heard from him yet regarding this PR. I'm happy to be patient, but I just want to make sure it is on someones todo list.

**Comment by 0xACE** [comment]
I've got no idea of how often he checks on here. But luckily git commits contains an e-mail address, you could e-mail him and check that way `git > github` ;-)

My guess would be that this is a rather large branch to test, requiring more effort than just merging it blindly :).

**Comment by johnnovak** [comment]
@JohnAD Sorry, I should check here more often. I'm a bit busy now but I will review this in a week or two. Seems like you've done some great work here, that's good! The screenshots of the tic-tac-toe game seem very nice!

So, I just want to reiterate, this library was originally meant for terminal output. I'm not against adding JS support, *as long as* it's not too invasive and does not overcomplicate the code and the API. Just be forewarned, I'll review the PR from this perspective 😄 Also, I don't want to cause any complications or additional complexities for users who simply don't care about supporting the browser. So, in my mind, the JS support should be a completely non-invasive "optional add-on".

But yeah, great work otherwise!

**Comment by johnnovak** [comment]
> My guess would be that this is a rather large branch to test, requiring more effort than just merging it blindly :)

I'm a bit afraid of this too, hence I put the disclaimer into my previous comment. We'll see how it goes.

The general direction seems to be good and that's how I would have done it (separating the backends into 3 files), so that's good so far 👍

**Comment by enthus1ast** [comment]
@JohnAD , was able to try it finally, really nice 👍

**Comment by johnnovak** [comment]
Ok, started reviewing. First of all, the output doesn't look quite alright in the browser on OS X. Tried both Firefox and Chrome, same results.

On Windows there's no weird white lines in the output.

Another comment: the white foreground on light blue (cyan?) background makes things hard to see in the Tic Tac Toe game. Can you please make it more contrasty? (e.g. by changing the foreground colour to black)

image

image

**Comment by johnnovak** [comment]
`forminput.nim` doesn't work in the browser:

image

Works in the terminal, but please use a better bg/fg colour combination for the input boxes, and get rid of the exit debug message.

image

File: README.md [hunk 1]
```diff
@@ -19,15 +19,16 @@ a full-blown GUI.
 ### Main features
 
 * **Non-blocking** keyboard input.
-* Support for **key combinations** and **special keys** available both in the standard
-  Windows Console (`cmd.exe`) and most common POSIX terminals.
+* Support for Windows Console (`cmd.exe`), POSIX (MacOS/Linux terminals), and 
+  Javascript (web broswer).
```
Comment (johnnovak) [inline README.md:23]: Typo: browser

Comment (johnnovak) [reply README.md:23]: Typo: JavaScript

File: README.md [hunk 2]
```diff
@@ -40,8 +41,8 @@ player that uses illwill for its awesome text-mode user interface</em></p>
 ### Use it if
 
 * You just want something simple that lets you write full-screen terminal
-  apps with non-blocking keyboard input that work on Windows, Mac OS X and
-  99.99% of modern Linux distributions.
+  apps with non-blocking keyboard input that work on Windows, Mac OS X, 
+  99.99% of modern Linux distributions, or any modern web browsers.
```
Comment (johnnovak) [inline README.md:45]: browser (singular)

File: README.md [hunk 3]
```diff
@@ -61,7 +62,7 @@ player that uses illwill for its awesome text-mode user interface</em></p>
 * Suspend/resume (SIGTSTP/SIGCONT handling) works, but it doesn't properly
   reset the terminal when suspending the app.
 * The contents of the terminal is not restored after exiting a full-screen app
-  on Windows.
+  on Windows or Javascript.
```
Comment (johnnovak) [inline README.md:65]: JavaScript

File: common.nim [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+    result="z".toRunes
+  of LeftBrace:
+    result="{".toRunes
+  of Pipe:
+    result="|".toRunes
+  of RightBrace:
+    result="}".toRunes
+  of Tilde:
+    result="~".toRunes
+  else:
```
Comment (johnnovak) [inline common.nim:522]: Minor: space around `=` (`result = ...`)

File: examples/badtictactoe.nim [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+  of gsEnd:
+    if tb.inputLineReady():
+      var answer = getInputLineText()
+      if answer == "new":
+        game.state = gsGameInit
+      else:
+        msgBox(tb, "Unknown command.")
+        tb.write(65, 21, bgBlue, "    ")
+        tb.startInputLine(65, 21, "", 4)
+        tb.display()
```
Comment (johnnovak) [inline examples/badtictactoe.nim:251]: Minor: blank lines between the cases would make things a bit easier to read.

File: examples/forminput.nim [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import illwill
+
+illwillInit(fullscreen=false)
+hideCursor()
+
+proc exitProc() {.noconv.} =
+  illwillDeinit()
+  echo "exit proc called"
```
Comment (johnnovak) [inline examples/forminput.nim:8]: Lose the debug echo

File: illwill.nim [hunk 7]
```diff
@@ -15,243 +16,65 @@
...
-## * `terminalWidth() <https://nim-lang.org/docs/terminal.html#terminalWidth>`_
-## * `terminalHeight() <https://nim-lang.org/docs/terminal.html#terminalHeight>`_
-## * `terminalSize() <https://nim-lang.org/docs/terminal.html#terminalSize>`_
-## * `hideCursor() <https://nim-lang.org/docs/terminal.html#hideCursor.t>`_
-## * `showCursor() <https://nim-lang.org/docs/terminal.html#showCursor.t>`_
-## * `Style <https://nim-lang.org/docs/terminal.html#Style>`_
-##
-
-import macros, os, terminal, unicode, bitops
+## When compiled for Javascript (JS), the `echo` command is safe to use but it's
```
Comment (johnnovak) [inline illwill.nim:25]: JavaScript

Comment (johnnovak) [reply illwill.nim:25]: Typo: its (not it's)

File: illwill.nim [hunk 8]
```diff
@@ -15,243 +16,65 @@
...
+## output is sent to the debug console NOT the terminal.
```
Comment (johnnovak) [inline illwill.nim:26]: Lowercase "not" will be fine :)

File: illwill.nim [hunk 9]
```diff
@@ -15,243 +16,65 @@
...
+export IllwillError
+export TerminalBuffer
+export TerminalChar
+export MouseButtonAction
+export MouseInfo
+export MouseButton
+export ScrollDirection
+
+proc terminalWidth*(): int =
+  ## From the context of the current terminal, returns the number of columns.
```
Comment (johnnovak) [inline illwill.nim:57]: Can you simplify this? "Returns the number of columns of the terminal"
Similarly for the other methods.

File: illwill.nim [hunk 10]
```diff
@@ -15,243 +16,65 @@
...
+  result = env_terminal.terminalWidth()
+
+proc terminalHeight*(): int =
+  ## From the context of the current terminal, returns the number of rows.
+  result = env_terminal.terminalHeight()
+
+proc terminalSize*(): tuple[w, h: int] =
+  result = env_terminal.terminalSize()
+  ## Returns the terminal width and height as a tuple. Internally calls
```
Comment (johnnovak) [inline illwill.nim:66]: I suggest dropping the second sentence, unnecessary detail that doesn't add much.

File: illwill.nim [hunk 11]
```diff
@@ -15,243 +16,65 @@
...
+  ## terminalWidth and terminalHeight, so the same assumptions apply.
+
+proc showCursor*() =
+  ## Shows the terminal cursor. In other words, the cursor can be seen
```
Comment (johnnovak) [inline illwill.nim:70]: Drop the second sentence.

File: illwill.nim [hunk 12]
```diff
@@ -15,243 +16,65 @@
...
+  ## at it's current position by end user.
+  env_terminal.showCursor()
+
+proc hideCursor*() =
+  ## Hides the terminal cursor. In other words, the cursor becomes hidden from
```
Comment (johnnovak) [inline illwill.nim:75]: Drop the second sentence.

File: illwill.nim [hunk 13]
```diff
@@ -714,30 +137,14 @@ proc exitFullScreen() =
...
-    )
-
-  proc disableMouse(hConsoleInput: Handle, oldConsoleMode: DWORD) =
-    discard setConsoleMode(hConsoleInput, oldConsoleMode) # TODO: REMOVE MOUSE OPTION ONLY?
-
 proc illwillInit*(fullScreen: bool = true, mouse: bool = false) =
   ## Initializes the terminal and enables non-blocking keyboard input. Needs
   ## to be called before doing anything with the library.
   ##
+  ## if fullScreen is set to true and the target is a POSIX terminal (MacOS,
```
Comment (johnnovak) [inline illwill.nim:144]: Typo: start sentence with uppercase

File: illwill.nim [hunk 14]
```diff
@@ -714,30 +137,14 @@ proc exitFullScreen() =
...
+  ## Linux, etc.), then the current screen is saved and will be restored
+  ## when illwillDeinit is invoked.
```
Comment (johnnovak) [inline illwill.nim:146]: `illwillDeinit()` (put it between backticks, add `()` at the end)

File: illwill.nim (hunk 13 above)
Comment (johnnovak) [reply illwill.nim:144]: Backticks around parameter names (`fullScreen`)

File: illwill.nim [hunk 15]
```diff
@@ -1023,16 +291,17 @@ proc newTerminalBufferFrom*(src: TerminalBuffer): TerminalBuffer =
   result = tb
 
 proc setCursorPos*(tb: var TerminalBuffer, x, y: Natural) =
-  ## Sets the current cursor position.
+  ## Sets the current cursor position. X is the column and Y is the row. Both
+  ## columns and rows start at zero (0).
```
Comment (johnnovak) [inline illwill.nim:295]: I think `(0)` can be removed from all comments, people know what zero is :)

File: illwill.nim [hunk 16]
```diff
@@ -1053,15 +322,15 @@ proc setStyle*(tb: var TerminalBuffer, style: set[Style]) =
...
   result = (tb.currX, tb.currY)
 
 func getCursorXPos*(tb: TerminalBuffer): Natural =
-  ## Returns the current x cursor position.
+  ## Returns the current x (column) cursor position.
   result = tb.currX
 
 func getCursorYPos*(tb: TerminalBuffer): Natural =
-  ## Returns the current y cursor position.
+  ## Returns the current y (row) cursor position.
```
Comment (johnnovak) [inline illwill.nim:333]: These comments are a bit wordy/noisy. I'd just state somewhere else once that the the top-left coordinate is (0,0) and be done with it...

File: illwill.nim [hunk 17]
```diff
@@ -1617,4 +895,230 @@ proc drawRect*(tb: var TerminalBuffer, x1, y1, x2, y2: Natural,
...
+  tb.write(bb)
+
+var
+  userInputText = ""
+  userInputActive = false
+  userInputExitKey = Key.None
+  userInputMaxLen = 1
+  userInputColumn = 0
+  userInputRow = 0
+  userInputAllowTabArrow = false
```
Comment (johnnovak) [inline illwill.nim:907]: These are globals, so please follow the existing convention of prefixing them with a `g`. So `gUserInput`, etc...

File: illwill.nim [hunk 18]
```diff
@@ -1617,4 +895,230 @@ proc drawRect*(tb: var TerminalBuffer, x1, y1, x2, y2: Natural,
...
+
+proc startInputLine*(
+                      tb: var TerminalBuffer, 
+                      x: int, 
+                      y: int, 
+                      strDefault: string, 
+                      strLen: int, 
+                      allowTabArrowExit = false
+                    ) =
```
Comment (johnnovak) [inline illwill.nim:916]: Plese use the existing indentation styles (this is not it).

File: illwill.nim [hunk 19]
```diff
@@ -1617,4 +895,230 @@ proc drawRect*(tb: var TerminalBuffer, x1, y1, x2, y2: Natural,
...
+  ## Setup the terminal for gathering text from the terminal user.
+  ##
+  ## The ``strDefault`` value is displayed at the column (x) and row(y). The
+  ## cursor is then moved to the end of that value and shown.
+  ##
+  ## If the column is too close the right edge fo the screen, then the length
+  ## is adjusted to account for that.
+  ##
+  ## This procedure does NOT actually gather then number. Instead, future calls
```
Comment (johnnovak) [inline illwill.nim:925]: Please no uppercase words, also something's wrong with this sentence ("gather then number"?).

File: illwill.nim [hunk 20]
```diff
@@ -1617,4 +895,230 @@ proc drawRect*(tb: var TerminalBuffer, x1, y1, x2, y2: Natural,
...
+  ## to ``inputLineReady`` will process future keypresses to build the string
+  ## and determine if the string is ready to be used.
+  ##
+  ## By default, only Enter (on Windows/POSIX) will finish the string entry (or
+  ## Ctrl-Enter on Javascript.) However, if ``allowTabArrowExit`` is set to true
+  ## then Escape, Up, Down, and Tab will also finish the entry.
```
Comment (johnnovak) [inline illwill.nim:931]: Ok, took me a while to realise what this is supposed to be about... I think you've implemented an input box with an optional default value and fixed length (with no scrolling), right? If so, can you make this whole description reflect that? It's not very clear right now.

File: illwill.nim [hunk 21]
```diff
@@ -1617,4 +895,230 @@ proc drawRect*(tb: var TerminalBuffer, x1, y1, x2, y2: Natural,
...
+                      strLen: int, 
+                      allowTabArrowExit = false
+                    ) =
+  ## Setup the terminal for gathering text from the terminal user.
+  ##
+  ## The ``strDefault`` value is displayed at the column (x) and row(y). The
+  ## cursor is then moved to the end of that value and shown.
+  ##
+  ## If the column is too close the right edge fo the screen, then the length
+  ## is adjusted to account for that.
```
Comment (johnnovak) [inline illwill.nim:923]: How does this handle window/terminal resizes? I think you might need to drop this feature of trimming the length to the available space.

File: illwill.nim [hunk 22]
```diff
@@ -1617,4 +895,230 @@ proc drawRect*(tb: var TerminalBuffer, x1, y1, x2, y2: Natural,
...
+  userInputRow = y
+  if strDefault.len > 0:
+    tb.write(strDefault)
+  userInputActive = true
+
+proc inputLineReady*(tb: var TerminalBuffer): bool =
+  ## This procedure both processes any incoming keys pressed (building the string),
+  ## but also returns a bool of true when the string is ready.
+  ##
+  ## When this value is true, any further processing is finished until ``startInputLine``
```
Comment (johnnovak) [inline illwill.nim:959]: `()` after methods please

File: illwill.nim [hunk 23]
```diff
@@ -1617,4 +895,230 @@ proc drawRect*(tb: var TerminalBuffer, x1, y1, x2, y2: Natural,
...
+  showCursor()
+  userInputColumn = goodX
+  userInputRow = y
+  if strDefault.len > 0:
+    tb.write(strDefault)
+  userInputActive = true
+
+proc inputLineReady*(tb: var TerminalBuffer): bool =
+  ## This procedure both processes any incoming keys pressed (building the string),
+  ## but also returns a bool of true when the string is ready.
```
Comment (johnnovak) [inline illwill.nim:957]: "returns true" - keep it simple

File: illwill.nim [hunk 24]
```diff
@@ -1617,4 +895,230 @@ proc drawRect*(tb: var TerminalBuffer, x1, y1, x2, y2: Natural,
...
+  tb.setCursorPos(goodX, y)
+  showCursor()
+  userInputColumn = goodX
+  userInputRow = y
+  if strDefault.len > 0:
+    tb.write(strDefault)
+  userInputActive = true
+
+proc inputLineReady*(tb: var TerminalBuffer): bool =
+  ## This procedure both processes any incoming keys pressed (building the string),
```
Comment (johnnovak) [inline illwill.nim:956]: "incoming keypresses"

File: illwill.nim [hunk 25]
```diff
@@ -1617,4 +895,230 @@ proc drawRect*(tb: var TerminalBuffer, x1, y1, x2, y2: Natural,
...
+  ## but also returns a bool of true when the string is ready.
+  ##
+  ## When this value is true, any further processing is finished until ``startInputLine``
+  ## restarts the whole process from the beginning again.
+  ##
+  ## If you wish to get the string, call ``getInputLineText``.
+  ##
+  ## If you wish to get the Key that caused the the process to stop, call
```
Comment (johnnovak) [inline illwill.nim:964]: `Key`

File: illwill.nim [hunk 26]
```diff
@@ -1617,4 +895,230 @@ proc drawRect*(tb: var TerminalBuffer, x1, y1, x2, y2: Natural,
...
+  ## restarts the whole process from the beginning again.
+  ##
+  ## If you wish to get the string, call ``getInputLineText``.
+  ##
+  ## If you wish to get the Key that caused the the process to stop, call 
+  ## ``getInputLineExitKey``.
+  if not userInputActive:
+    result = true
+    return
+  result = false
```
Comment (johnnovak) [inline illwill.nim:969]: Just do `return true`

File: illwill.nim [hunk 27]
```diff
@@ -1617,4 +895,230 @@ proc drawRect*(tb: var TerminalBuffer, x1, y1, x2, y2: Natural,
...
+        tb.write($r)
+        userInputText &= $r
+        changeMade = true
+  #
+  if userInputActive:
+    tb.setCursorPos(userInputColumn + userInputText.len, userInputRow)
+  else:
+    hideCursor()
+  if changeMade:
+    tb.display()
```
Comment (johnnovak) [inline illwill.nim:1024]: I don't like it that you're also updating the buffer in a piecewise fashion from this method. Please separate the two: just update the string buffer here, then somewhere else update the terminal buffer with the updated string. That's cleaner and performance is a complete non-issue here.

File: illwill.nim [hunk 28]
```diff
@@ -1617,4 +895,230 @@ proc drawRect*(tb: var TerminalBuffer, x1, y1, x2, y2: Natural,
...
+
+proc getInputLineText*(): string =
+  ## Get the current string being built (or finished being built) by the
+  ## ``inputLineReady`` procedure and started by the ``startInputLine`` procedure.
+  result = userInputText
+
+proc getInputLineExitKey*(): Key =
+  ## Get the reason that the ``startInputLine`` / ``inputLineReady`` process
```
Comment (johnnovak) [inline illwill.nim:1032]: "reason why"

File: illwill.nim [hunk 29]
```diff
@@ -1617,4 +895,230 @@ proc drawRect*(tb: var TerminalBuffer, x1, y1, x2, y2: Natural,
...
+  ## ended. If the process has not ended (or was never started), you will get a
+  ## value of Key.None.
```
Comment (johnnovak) [inline illwill.nim:1034]: `Key.None`

File: illwill.nim [hunk 30]
```diff
@@ -1617,4 +895,230 @@ proc drawRect*(tb: var TerminalBuffer, x1, y1, x2, y2: Natural,
...
+  result = userInputExitKey
+
+
+template timerLoop*(msDelay: int, content: untyped): untyped =
+  ## Loop through the subtending code every ``msDelay`` milliseconds. It is used
```
Comment (johnnovak) [inline illwill.nim:1039]: "Loop through the **passed in code block**"

File: illwill.nim [hunk 31]
```diff
@@ -1617,4 +895,230 @@ proc drawRect*(tb: var TerminalBuffer, x1, y1, x2, y2: Natural,
...
+  ##       , 20.0)
+  ##
+  ## The library detect the current compilation target (``c`` for Windows,
+  ## MacOS, Linux or ``js`` for Javascript) and invokes the correct version of
+  ## ``timerLoop``.
+  env_terminal.timerLoop(msDelay, content)
+
+
+proc illwillOnLoad*(domId: cstring, cols, rows: int) =
+  ## When using *illwill* to make Javascript, this is the
```
Comment (johnnovak) [inline illwill.nim:1094]: "When using the JavaScript backend"

File: js_terminal.nim [hunk 32]
```diff
@@ -0,0 +1,341 @@
...
+
+var
+  cursorX = 0
+  cursorY = 0
+  cursorOn = false
+  cursorBlockVisible = false 
+
+var
+  globalTimer*: ref Interval
+  cursorTimer*: ref Interval
```
Comment (johnnovak) [inline js_terminal.nim:87]: As mentioned before, all globals should be prefixed by `g`.

File: js_terminal.nim [hunk 33]
```diff
@@ -0,0 +1,341 @@
...
+  cursorX = x
+
+
+proc enableMouse*() =
+  discard
+
+
+proc disableMouse*() =
+  discard
+
```
Comment (johnnovak) [inline js_terminal.nim:148]: A single new line between these procs will suffice.

File: illwill.nimble [hunk 34]
```diff
@@ -32,5 +32,6 @@ task examplesDebug, "Compiles the examples (debug mode)":
   exec "nim c examples/mouseMinimal.nim"
 
 task docgen, "Generate HTML documentation":
-  exec "nim doc -o:doc/illwill.html illwill"
+  exec "nim doc --d:docgenonly -o:doc/illwill.html illwill.nim"
+  exec "nim doc --d:docgenonly -o:doc/common.html common.nim"
```
Comment (johnnovak) [inline illwill.nimble:36]: The nimble file needs some more work:

- increment version number
- add the new examples to the existing build targets
- add a new JS target to compile just the new JS example for the browser

File: illwill.nim [hunk 35]
```diff
@@ -1617,4 +895,230 @@ proc drawRect*(tb: var TerminalBuffer, x1, y1, x2, y2: Natural,
...
+  result = userInputText
+
+proc getInputLineExitKey*(): Key =
+  ## Get the reason that the ``startInputLine`` / ``inputLineReady`` process
+  ## ended. If the process has not ended (or was never started), you will get a
+  ## value of Key.None.
+  result = userInputExitKey
+
+
+template timerLoop*(msDelay: int, content: untyped): untyped =
```
Comment (johnnovak) [inline illwill.nim:1038]: I'd rename this to `eventLoop`, `msDelay` to `pollInterval` and `content` to `block`.

**Comment by johnnovak** [comment]
1. `timerLoop` (which should be `eventLoop`) doesn't handle resizing the terminal window. Run some of the existing examples and resize the terminal window and see what happens, then repeat with the new examples that use `timerLoop`.

2. styleddisplay: Clean exit must be implemented, just like in the existing examples.

3. I'm not a fan of dumping all the source files into the root folder. `illwill.nim` can stay there, but please put the rest into `src` or something. Make sure you adhere to the `nimble` packaging standard and verify with `nimble check`.

File: illwill.nim (hunk 18 above)
Comment (johnnovak) [reply illwill.nim:916]: Rename `strDefault` to `defaultValue`
Rename `strLen` to `maxLen`

File: illwill.nim (hunk 20 above)
Comment (johnnovak) [reply illwill.nim:931]: The list of exit keys seems arbitrary; because you are putting this into a generic library you need to make it generic enough so it's flexible enough for different use cases. I'd change this so it accepts a sequence of exit keys instead of just a flag (no need to validate it, if the user passes in stupid stuff like the keys A, B and C, it's their problem...)

File: js_terminal.nim [hunk 36]
```diff
@@ -0,0 +1,341 @@
...
+proc consoleDeInit*() =
+  window.clearInterval(globalTimer)
+  window.clearInterval(cursorTimer)
+  window.removeEventListener(
+    "keydown".cstring, 
+    onkeydown
+  )
+  if preTag.isNil:
+    echo "cdi not loaded yet"
```
Comment (johnnovak) [inline js_terminal.nim:322]: Can we provide a more useful message? What is `cdi`, is the user supposed to know that?

**Review by johnnovak — CHANGES_REQUESTED** [review]

**Comment by johnnovak** [comment]
@JohnAD First of all, thanks for this, there's lots of things that I like in your work. For example, separating the backends into different files was a good idea and it has been long overdue. I also applaud the effort to create some kind of a simple text input box.

I'm not totally conviced that adding a JS backend is useful, though. It complicates the API, e.g. you needed to introduce `timerLoop` to make it work. Realistically, who would want to write a console app and then want it to run it in the browser too? Seems very unlikely to me. Can you think of any use cases? I can't, because I personally would rather write a backend that contains just the processing, logic, algorithms and whatnot, then different frontends for it: one for the console using illwill, a completely new browser based UI, then a completely different graphical UI, etc. You know, probably no one would want to simulate the console in OpenGL or something either--they would just write a brand new graphical UI that's much better for their app.

I'm personally not interested in maintaining the JS backend at all, and unless there's a very strong argument for including it, I don't think I want to get it merged in.

Either way, I think the rest of the work minus the JS backend should probably go in.

Anyway, here's a list of further notes and comments, from which you can already see there's a number of issues introduced by `timerLoop`, which was necessitated by the JS backend. And we can't run the existing examples in the browser either without further modifications...

1. `timerLoop` (which should be `eventLoop`) doesn't handle resizing the terminal window. Run some of the existing examples and resize the terminal window and see what happens, then repeat with the new examples that use `timerLoop`. Basically, you should just create a new terminal buffer in each loop iteration in `timerLoop` with the current dimensions of the terminal, and then redraw everything into it.

1. badtictactoe: You're kind of assuming a 90x24 terminal window, but you can't make such assumptions. At the very least, you should make sure that if the terminal window is smaller, the output is clipped to the visible area. For bonus points you could detect that the terminal is smaller than 90x24 and display a message like "This example needs a 90x24 terminal window (or larger), please resize your terminal window."

1. styleddisplay: Clean exit must be implemented, just like in the existing examples.

1. I'm not a fan of dumping all the source files into the root folder. Then you have `illwill.nim`, and `common.nim` in addition to that. I suggest the following structure as the current best-practice:

```
illwill.nim     - contains just the exports, lots of stuff can be moved into common
illwill/
  common.nim
  js.nim        - just lose _terminal from the backend names
  posix.nim
  windows.nim

```

5. There's lots of typos and sentences with strange wording in the doco.  I gave up commenting on all of these after a while, it would be easier to just rewrite it myself...  Please read through everything carefully a few times and try to improve it. Also please try to be succint.

**Comment by JohnAD** [comment]
Thanks for all the feedback! I'm hoping to work on it again next weekend.

**Comment by codewski** [comment]
@JohnAD @johnnovak  is this gonna merge or is abandoned? I'd like to use illwill for a project. Thanks in advance.

**Comment by JohnAD** [comment]
@codewski Unfortunately, I've gotten far busier, so I'm not sure when I'll be able to work on it again. Ironically, it is not far from finished.

Myself, I'm using my local fork for a few projects.

Feel free to fork my repo: https://github.com/JohnAD/illwill

And, if you end up finishing it, you can either make a new PR to merge it with Illwill. Or you can rename it into a new library for nimble. johnnovak was not looking forward to supporting JS output, so I suspect he would be okay with a js-oriented version based on his workk.

**Comment by johnnovak** [comment]
> johnnovak was not looking forward to supporting JS output, so I suspect he would be okay with a js-oriented version based on his work.

That's right, this is meant to be used in "proper" terminals, so JS output won't be added as it complicates everything a bit too much for my taste for such a niche use-case. But of course, feel free to fork it and do whatever 😄
